File: src/data/avatars.ts
import { AvatarKind, type PlayerAvatar } from '../types/game';

export const PRESET_AVATARS = {
  Cat: 'cat',
  Dog: 'dog',
  Robot: 'robot',
  Alien: 'alien',
  Ghost: 'ghost',
  Penguin: 'penguin',
  Fox: 'fox',
  Dino: 'dino',
  Crab: 'crab',
  Octopus: 'octopus',
} as const;
export type PresetAvatarId = (typeof PRESET_AVATARS)[keyof typeof PRESET_AVATARS];

export const PRESET_EMOJI: Record<PresetAvatarId, string> = {
  [PRESET_AVATARS.Cat]: '🐱',
  [PRESET_AVATARS.Dog]: '🐶',
  [PRESET_AVATARS.Robot]: '🤖',
  [PRESET_AVATARS.Alien]: '👽',
  [PRESET_AVATARS.Ghost]: '👻',
  [PRESET_AVATARS.Penguin]: '🐧',
  [PRESET_AVATARS.Fox]: '🦊',
  [PRESET_AVATARS.Dino]: '🦖',
  [PRESET_AVATARS.Crab]: '🦀',
  [PRESET_AVATARS.Octopus]: '🐙',
};

export const DEFAULT_AVATAR: PlayerAvatar = { kind: AvatarKind.Preset, id: PRESET_AVATARS.Cat };

export const CUSTOM_AVATAR_MAX_DATA_URL_LENGTH = 100_000;
export const CUSTOM_AVATAR_MAX_DIMENSION = 96;

export function isPresetAvatar(value: unknown): value is { kind: typeof AvatarKind.Preset; id: PresetAvatarId } {
  if (typeof value !== 'object' || value === null) return false;
  const v = value as Record<string, unknown>;
  return v.kind === AvatarKind.Preset && typeof v.id === 'string' && Object.hasOwn(PRESET_EMOJI, v.id);
}

export function isCustomAvatar(value: unknown): value is { kind: typeof AvatarKind.Custom; dataUrl: string } {
  if (typeof value !== 'object' || value === null) return false;
  const v = value as Record<string, unknown>;
  if (v.kind !== AvatarKind.Custom) return false;
  if (typeof v.dataUrl !== 'string') return false;
  if (v.dataUrl.length > CUSTOM_AVATAR_MAX_DATA_URL_LENGTH) return false;
  return v.dataUrl.startsWith('data:image/');
}

export function isValidAvatar(value: unknown): value is PlayerAvatar {
  return isPresetAvatar(value) || isCustomAvatar(value);
}

export function avatarEmoji(avatar: PlayerAvatar): string | null {
  if (avatar.kind !== AvatarKind.Preset) return null;
  return Object.hasOwn(PRESET_EMOJI, avatar.id) ? (PRESET_EMOJI[avatar.id] ?? null) : null;
}

File: backend/internal/data/avatars.go
package data

import (
	"strings"

	"monopoly-game-backend/internal/entity"
)

var PRESET_AVATARS = map[string]string{
	"Cat":     "cat",
	"Dog":     "dog",
	"Robot":   "robot",
	"Alien":   "alien",
	"Ghost":   "ghost",
	"Penguin": "penguin",
	"Fox":     "fox",
	"Dino":    "dino",
	"Crab":    "crab",
	"Octopus": "octopus",
}

var PRESET_EMOJI = map[string]string{
	"cat":     "\U0001F431",
	"dog":     "\U0001F436",
	"robot":   "\U0001F916",
	"alien":   "\U0001F47D",
	"ghost":   "\U0001F47B",
	"penguin": "\U0001F427",
	"fox":     "\U0001F98A",
	"dino":    "\U0001F996",
	"crab":    "\U0001F980",
	"octopus": "\U0001F419",
}

var DEFAULT_AVATAR entity.PlayerAvatar = entity.PresetAvatar{
	Kind: entity.AvatarKindPreset,
	ID:   "cat",
}

const CUSTOM_AVATAR_MAX_DATA_URL_LENGTH = 100_000
const CUSTOM_AVATAR_MAX_DIMENSION = 96

func IsPresetAvatar(value entity.PlayerAvatar) bool {
	pa, ok := value.(entity.PresetAvatar)
	if !ok {
		return false
	}
	if pa.Kind != entity.AvatarKindPreset {
		return false
	}
	_, exists := PRESET_EMOJI[pa.ID]
	return exists
}

func IsCustomAvatar(value entity.PlayerAvatar) bool {
	ca, ok := value.(entity.CustomAvatar)
	if !ok {
		return false
	}
	if ca.Kind != entity.AvatarKindCustom {
		return false
	}
	if len(ca.DataURL) > CUSTOM_AVATAR_MAX_DATA_URL_LENGTH {
		return false
	}
	return strings.HasPrefix(ca.DataURL, "data:image/")
}

func IsValidAvatar(value entity.PlayerAvatar) bool {
	return IsPresetAvatar(value) || IsCustomAvatar(value)
}

func IsSameAvatar(a, b entity.PlayerAvatar) bool {
	aPreset, aOk := a.(entity.PresetAvatar)
	bPreset, bOk := b.(entity.PresetAvatar)
	if aOk && bOk {
		return aPreset.Kind == bPreset.Kind && aPreset.ID == bPreset.ID
	}

	aCustom, aOk := a.(entity.CustomAvatar)
	bCustom, bOk := b.(entity.CustomAvatar)
	if aOk && bOk {
		return aCustom.Kind == bCustom.Kind && aCustom.DataURL == bCustom.DataURL
	}

	return false
}
